### ISSUE TITLE
Do not watch unsupported file systems

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#15881]
File system watching might be problematic when it comes to "exotic" file systems like FAT*, SMB and NFS. By "exotic" here we mean any file system that is not the "native" non-network FS of the operating system, i.e. NTFS for Windows, ext3/ext4 for Linux and APFS/HFS+ for macOS.

We currently warn the user (in a rather obscure place) in the user guide that (some of) these file systems are not supported for file system watching. However, what "not supported" means is not very clear. Indeed, we don't have any tests on CI with these use cases, so behavior is not well defined.

This was acceptable for Gradle 6.7 and 6.8 where file system watching was opt-in, however in Gradle 7.0 we want to enable by default. This might cause builds involving exotic FSs to miss notifications about file changes and produce incorrect results without the user noticing. This should be avoided.

Considering that we (currently) watch only the project root and whatever's underneath it, the exotic FS can appear in different ways:

1) The project root is on the exotic FS
2) A directory under the project root is used to mount an exotic FS
3) A symlink under the project root links to content on an exotic FS
4) The configuration of the build references an exotic FS directly (e.g. as a task's input)

Open questions:

- We basically talk about FAT on Windows, and assume that all other non-network file systems used in a build are the system-native ones (like ext3/4 for Linux, NTFS for Windows and HFS+/APFS for macOS). However, this is a somewhat oversimplified view:
  - there are many forms of FAT: FAT16, FAT32, exFAT etc. See https://en.wikipedia.org/wiki/File_Allocation_Table#Types
  - FAT and other file systems like btrfs, tmpfs and zfs are available on Linux and macOS to mount
  - NTFS can be mounted read-only on *NIX-es
  - there's stuff 
- Network file systems have an underlying file system which might produce different behavior on the same network protocol.
- What about WSL and WSL2 on Windows?

---
cc: @gradle/execution

**Comment by lptr** [comment]
For 1) we need to do further testing. Some notes:

- Running a build from a project root on a FAT file system on Windows is possible.
- With SMB you need to map the network volume to a drive, because you running Gradle directly from `\\server\...` doesn't work (we fail with an exception).
- Running a build from NFS might not work because of some locking problems, we need to check this.

It would be ideal if we could disable file system watching completely if the project was built from a project root on an exotic FS.

For 2) we should mention in the user guide that this use case is not supported and behavior is undefined. We should watch for community feedback whether this becomes a problem, and then address it if we have to.

For 3) we don't have to do anything as we already ignore content accessed via a symlink.

For 4) we also don't need to do anything, as we are not watching things outside the project root.

**Comment by lptr** [comment]
After testing a number of file systems it seems that the ones we could think of work fine (`ext3`, `ext4`, `ntfs`, `apfs`, `hfs+`, `btrfs`, `fat32`) on all the operating systems we tested.

What doesn't work is networked file systems. It seems that while Samba in theory is capable of transmitting file system events, it does not do it between any of the host/client OS we tested. To clarify: changes initiated on the client computer do produce notifications, but changes initiated on the host (remote) computer do not.

See internal document about the investigation here: https://docs.google.com/document/d/1ByfSIU_81-4UXcKiMeInWjV5IWZSl7lG39FCX5NQSY8

Unfortunately this will break any build out there that relies on changes being tracked on the host OS with file system watching being enabled by default.

Luckily, `native-platform`'s `FileSystems` service has just what we need:

```groovy
import net.rubygrapefruit.platform.Native
import net.rubygrapefruit.platform.file.FileSystems

Native.get(FileSystems.class).fileSystems.each { fs ->
    println "${fs.fileSystemType} - ${fs.mountPoint} - ${fs.deviceName} - ${fs.remote}"
}
```

This produces the following on macOS:

```text
apfs - / - /dev/disk1s1s1 - false
devfs - /dev - devfs - false
apfs - /System/Volumes/VM - /dev/disk1s5 - false
apfs - /System/Volumes/Preboot - /dev/disk1s3 - false
apfs - /System/Volumes/Update - /dev/disk1s6 - false
apfs - /System/Volumes/Data - /dev/disk1s2 - false
ntfs - /Volumes/DATA - /dev/disk0s3 - false
ntfs - /Volumes/BOOTCAMP - /dev/disk0s4 - false
autofs - /System/Volumes/Data/home - map auto_home - true
nullfs - /private/var/folders/bl/_sc1qdv14v7d861g20drsf7r0000gn/T/AppTranslocation/7B556C63-8713-40BE-BC44-1CAB64AC7B08 - /Applications/Original Prusa Drivers/PrusaSlicer.app - false
smbfs - /Users/lptr/Workspace/gradle/test/mount - //lptr@lorant-mbp/fsw-test - true
```

...and this on Windows:

```text
NTFS - C:\ - \Device\HarddiskVolume3 - false
NTFS - N:\ - \Device\LanmanRedirector;N:0000000000022502\NASI\fsw-test - true
NTFS - S:\ - ??\N: - true
NTFS - T:\ - ??\C: - false
```

Note that

- `/Users/lptr/Workspace/gradle/test/mount` is a SMB mount from a Windows machine,
- `N:` is an SMB mount from a Linux machine,
- `S:` is a `SUBST` of `N:\` (correctly marked as remote)
- `T:` is a `SUBST` of `C:\` (correctly marked as not remote)

Basically, our known supported list of file systems looks like this:

- `apfs`, `hfs+`, `btrfs`, `ext3`, `ext4`, `ntfs`

We also require the file system not to be remote.

Gradle should behave thusly:

- No watching flag provided: enable for the above known supported file systems only
- Watching flag is `true`: enable regardless of file system type
- Watching flag is `false`: disable watching

**Comment by lptr** [comment]
Remaining issues:

- Check each time we start watching a path whether it's on a supported file system if FSW is enabled by default
  - With Linux doing the check with every directory might be too expensive, so we might just do it once per hierarchy, and ignore potential mount points hidden in the hiearchy.
- Collect actual list of supported file systems (we need to check `fat32` etc.), though if we miss something, we can always enable it in a later release.
- Add tests (this will be hard).
- Document all this.

**Comment by wolfs** [comment]
We should probably add the in-memory file systems to the list of supported file systems. Things like `tmpfs` and `ramfs`. And Docker also uses a certain file system in their containers - we should add that too.

**Comment by lptr** [comment]
Note on macOS if you create a ramdisk like this, it will have `hfs` as the file system:

https://gist.github.com/htr3n/344f06ba2bb20b1056d7d5570fe7f596

**Comment by lptr** [comment]
macOS's `diskutil` supports the following file systems:

```console
$ diskutil listfilesystems
Formattable file systems

These file system personalities can be used for erasing and partitioning.
When specifying a personality as a parameter to a verb, case is not considered.
Certain common aliases (also case-insensitive) are listed below as well.

-------------------------------------------------------------------------------
PERSONALITY                     USER VISIBLE NAME
-------------------------------------------------------------------------------
Case-sensitive APFS             APFS (Case-sensitive)
  (or) APFSX
APFS                            APFS
  (or) APFSI
ExFAT                           ExFAT
Free Space                      Free Space
  (or) FREE
MS-DOS                          MS-DOS (FAT)
MS-DOS FAT12                    MS-DOS (FAT12)
MS-DOS FAT16                    MS-DOS (FAT16)
MS-DOS FAT32                    MS-DOS (FAT32)
  (or) FAT32
HFS+                            Mac OS Extended
Case-sensitive HFS+             Mac OS Extended (Case-sensitive)
  (or) HFSX
Case-sensitive Journaled HFS+   Mac OS Extended (Case-sensitive, Journaled)
  (or) JHFSX
Journaled HFS+                  Mac OS Extended (Journaled)
  (or) JHFS+
```

**Comment by abueide** [comment]
@lptr is zfs supported in 7.0? Currently on my NixOS machine I'm running zfs & getting this error on my gradle build which i've never seen on my other machines.

https://gist.github.com/abueide/335676ef82cdbb46c530952aad3deccb

**Comment by lptr** [comment]
@abueide that is not intended behavior. Can you please open a separate issue?

**Comment by AbelAlejandro** [comment]
I'm having problems with this on Gradle 6.7 and JDK 8 on Gitlab CI Pipelines. I'm not sure what file system is `gradle:6.7-jdk8` using but I didn't expect it to be an 'exotic' one.

https://gist.github.com/AbelAlejandro/2ee20f826f21b63018125fa622f64e0a

**Comment by lptr** [comment]
@AbelAlejandro does this happen with 6.7.1 as well? We fixed https://github.com/gradle/gradle/issues/15086 that seems relevant.

**Comment by lptr** [comment]
Actually, the stack trace shows `DefaultWatchableFileSystemDetector` which was introduced in Gradle 7.0, so it looks like you are using that version. Upgrading to 7.0.1 should fix the problem, see: https://github.com/gradle/gradle/pull/16899.

**Comment by AbelAlejandro** [comment]
Thanks a lot. I was using the gradle wrapper instead of the gradle from the image. My bad. Running on 6.7.1 works fine.
